includes/class-omise-callback.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * @since 4.0
+ */
+class Omise_Callback {
+	/**
+	 * @var \WC_Abstract_Order
+	 */
+	protected $order;
+
+	/**
+	 * @var \OmiseCharge
+	 */
+	protected $charge;
+
+	/**
+	 * @param \WC_Abstract_Order $order
+	 */
+	public function __construct( $order ) {
+		$this->order = $order;
+		if ( ! $this->order || ! $this->order instanceof WC_Abstract_Order ) $this->invalid_result();
+	}
+
+	public static function execute() {
+		$order_id = isset( $_GET['order_id'] ) ? sanitize_text_field( $_GET['order_id'] ) : null;
+
+		$callback = new self( wc_get_order( $order_id ) );
+		$callback->validate();
+	}
+
+	public function validate() {
+		$this->order->add_order_note( __( 'OMISE: Validating the payment result...', 'omise' ) );
+
+		try {
+			$this->charge = OmiseCharge::retrieve( $this->order->get_transaction_id() );
+
+			switch ( strtolower( $this->charge['status'] ) ) {
+				case 'successful':
+				case 'failed':
+				case 'pending':
+					$resolving_method = strtolower( 'payment_' . $this->charge['status'] );
+					$this->$resolving_method();
+					break;
```

Comment:
I still didn't change this part as this helps organising the code better to me, comparing to 
```php
switch ( strtolower( $this->charge['status'] ) ) {
    case 'successful':
        $this->payment_successful();
        break;

    case 'failed':
        $this->payment_failed();
        break;

    case 'pending':
        $this->payment_pending();
        break;

...
```
But if everyone strongly suggests the above ☝️ over the current, then I'll give up 😂

Reply:
It seems fine to me

Reply:
How about moving the switch into its own `dispatch` function if you think it looks messy.

```php
$this->dispatch($this->charge['status'])
```

Also allows you to `$callback->dispatch('successful')` from the outside if you need that.

Reply:
Thanks for suggesting 👍 
I have just write some part of it, the dispatch idea sounds interesting. But probably I'll leave it as it is for now until the requirement of using `dispatch()` outside of Callback context comes.  

```php
public function validate() {
    $this->order->add_order_note( __( 'OMISE: Validating the payment result...', 'omise' ) );

    try {
        $this->charge = OmiseCharge::retrieve( $this->order->get_transaction_id() );
        $this->dispatch( $this->charge['status'] );
    } catch ( Exception $e ) {
        $this->invalid_result( $e->getMessage() );
    }
}

public function dispatch( $status ) {
    if ( ! in_array( $status, array( 'successful', 'failed', 'pending' ) ) ) {
        $this->invalid_result( __( 'Unrecognized Omise Charge status.', 'omise' ) );
    }

    $resolving_method = strtolower( 'payment_' . $this->charge['status'] );
    $this->$resolving_method();
}

/**
 * Resolving a case of undefined charge status
 */
protected function invalid_result( $order_note = null ) {
    if ( $order_note ) {
        $this->order->add_order_note(
            sprintf(
                wp_kses( __( 'OMISE: Unable to validate the result.<br/>%s', 'omise' ), array( 'br' => array() ) ),
                $order_note
            )
        );
    }

    $message = __(
        '<strong>We cannot validate your payment result:</strong><br/>
            Note that your payment may have already been processed.<br/>
            Please contact our support team if you have any questions.',
        'omise'
    );

    wc_add_notice( wp_kses( $message, array( 'br' => array(), 'strong' => array() ) ), 'error' );
    wp_redirect( wc_get_checkout_url() );
    exit;
}
```